File: starlore-py/app/services/ai_quota_service.py
"""AI 配额管理服务。"""

from datetime import date

from sqlalchemy import select, update
from sqlalchemy.ext.asyncio import AsyncSession

from app.models.user import User

# 角色默认每日额度：admin 不限流，member 99 次，其余角色 10 次
_ROLE_DEFAULT_LIMITS = {"member": 99}
_DEFAULT_LIMIT = 10


def _effective_limit(user: User) -> int:
    """生效额度：用户个人 ai_daily_limit 优先，未设置时按角色默认。"""
    return user.ai_daily_limit or _ROLE_DEFAULT_LIMITS.get(user.role, _DEFAULT_LIMIT)


async def get_remaining(db: AsyncSession, user_id: int) -> int:
    """获取用户今日剩余 AI 调用次数。-1 表示无限制。"""
    result = await db.execute(select(User).where(User.id == user_id))
    user = result.scalar_one_or_none()
    if user is None:
        return 0
    if user.role == "admin":
        return -1

    # 日期重置
    if user.ai_reset_date != date.today():
        user.ai_today_count = 0
        user.ai_reset_date = date.today()
        await db.flush()

    limit = _effective_limit(user)
    return max(0, limit - user.ai_today_count)


async def try_consume(db: AsyncSession, user_id: int) -> bool:
    """尝试消耗一次配额。成功返回 True，配额耗尽返回 False。"""
    result = await db.execute(select(User).where(User.id == user_id))
    user = result.scalar_one_or_none()
    if user is None:
        return False
    if user.role == "admin":
        return True

    # 日期重置
    if user.ai_reset_date != date.today():
        user.ai_today_count = 0
        user.ai_reset_date = date.today()

    limit = _effective_limit(user)
    if user.ai_today_count >= limit:
        return False

    user.ai_today_count += 1
    await db.flush()
    return True


async def get_quota_info(db: AsyncSession, user_id: int) -> dict:
    """获取配额详情。"""
    result = await db.execute(select(User).where(User.id == user_id))
    user = result.scalar_one_or_none()
    if user is None:
        return {"dailyLimit": 10, "used": 0, "remaining": 0, "isAdmin": False}

    is_admin = user.role == "admin"

    # 日期重置
    if user.ai_reset_date != date.today():
        user.ai_today_count = 0
        user.ai_reset_date = date.today()
        await db.flush()

    limit = _effective_limit(user)
    used = user.ai_today_count

    return {
        "dailyLimit": limit,
        "used": used,
        "remaining": -1 if is_admin else max(0, limit - used),
        "isAdmin": is_admin,
    }

File: starlore-back/src/main/java/com/robin/blogback/service/AiQuotaService.java
package com.robin.blogback.service;

import com.baomidou.mybatisplus.core.conditions.update.LambdaUpdateWrapper;
import com.robin.blogback.entity.User;
import com.robin.blogback.mapper.UserMapper;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.time.LocalDate;

@Service
public class AiQuotaService {

    @Autowired
    private UserMapper userMapper;

    /** 获取用户今日 AI 剩余次数（返回 -1 表示不限） */
    public int getRemaining(Integer userId) {
        if (userId == null) return 0;
        User user = userMapper.selectById(userId);
        if (user == null) return 0;

        // 管理员不限
        if ("admin".equals(user.getRole())) return -1;

        // 检查是否需要重置今日计数
        LocalDate today = LocalDate.now();
        if (!today.equals(user.getAiResetDate())) {
            userMapper.update(null, new LambdaUpdateWrapper<User>()
                    .eq(User::getId, userId)
                    .set(User::getAiTodayCount, 0)
                    .set(User::getAiResetDate, today));
            return user.getAiDailyLimit() != null ? user.getAiDailyLimit() : 10;
        }

        int limit = user.getAiDailyLimit() != null ? user.getAiDailyLimit() : 10;
        int used = user.getAiTodayCount() != null ? user.getAiTodayCount() : 0;
        return Math.max(0, limit - used);
    }

    /** 尝试消耗一次配额，返回 true=成功，false=配额不足 */
    public boolean tryConsume(Integer userId) {
        if (userId == null) return false;
        User user = userMapper.selectById(userId);
        if (user == null) return false;

        // 管理员不限
        if ("admin".equals(user.getRole())) return true;

        LocalDate today = LocalDate.now();
        int limit = user.getAiDailyLimit() != null ? user.getAiDailyLimit() : 10;

        // 跨天重置
        if (!today.equals(user.getAiResetDate())) {
            userMapper.update(null, new LambdaUpdateWrapper<User>()
                    .eq(User::getId, userId)
                    .set(User::getAiTodayCount, 1)
                    .set(User::getAiResetDate, today));
            return true;
        }

        int used = user.getAiTodayCount() != null ? user.getAiTodayCount() : 0;
        if (used >= limit) return false;

        userMapper.update(null, new LambdaUpdateWrapper<User>()
                .eq(User::getId, userId)
                .setSql("ai_today_count = ai_today_count + 1"));
        return true;
    }

    /** 获取用户配额详情 */
    public QuotaInfo getQuotaInfo(Integer userId) {
        if (userId == null) return new QuotaInfo(0, 0, 0, false);
        User user = userMapper.selectById(userId);
        if (user == null) return new QuotaInfo(0, 0, 0, false);

        LocalDate today = LocalDate.now();
        int limit = user.getAiDailyLimit() != null ? user.getAiDailyLimit() : 10;
        int used = user.getAiTodayCount() != null ? user.getAiTodayCount() : 0;

        // 跨天重置
        if (!today.equals(user.getAiResetDate())) {
            used = 0;
        }

        boolean isAdmin = "admin".equals(user.getRole());
        return new QuotaInfo(isAdmin ? -1 : limit, used,
                isAdmin ? -1 : Math.max(0, limit - used), isAdmin);
    }

    public record QuotaInfo(int dailyLimit, int used, int remaining, boolean isAdmin) {}
}
